[WIP] Sync bias

jsirex opened the pull request:
Do not merge yet.
Discussion here or https://github.com/itgmania/itgmania/discussions/268

I even don't know will I able to finish it. Not a programmer.

- [x] Machine Option SyncBias
- [x] TimingData SyncBias
- [x] NotesLoader SSC
- [x] NotesWriter SSC
- [ ] Bump Notes Version (skip?)
- [x] Screen Editor
- [x] Calculation Logic
- [x] Simply Love System Option: SyncBias
- [ ] AutoSync updates machine's SyncBias
- [ ] Cache issue: FIRSTSECOND/LASTSECOND have absolute value after offset correction. Need to think
- [ ] Assist tick strange behavior: some times it goes with notes, sometimes it goes with music

jsirex committed:
Add machine SyncBias preference

jsirex committed:
Add SyncBias to TimingData and Notes

jsirex committed:
Add Sync Bias to ScreenEdit

teejusb commented:
Personally I would wait to work on this because I do have some opinions that I haven't had time to bring up (working on other higher priority things atm for next release).

jsirex committed:
Add SyncBias logic

- if song or machine have undefined SyncBias preserve old behavior
- complatible with previous version synced and unsynced songs
- Implemented only for SSC type format

jsirex commented:
This is the hardest, complicated code I've ever seen. IDK, can I fix this..

Long read:

# General principle here: smooth upgrade
If it is an old version of ITGmania - nothing changes
If it is an old version of simfiles - nothing changes
If it is a new version of ITGmania:
- if machine's sync bias or song's sync bias is N/A - nothing changes
- if both sync bias set - appropriate calculation is happen

It gives opportunity to mark existing songs with `SyncBias` tag.
Any machine which has explicitly synced to Null or ITG (system settings is not N/A) will properly calculate offsets both for ITG and Null.

# Use cases
This is important change and requires proper testing. 
Let brute force all possibilities.

We can have machine configured:

1. SyncBias=N/A - Actually Null (Machine was synced to a Null Song)
2. SyncBias=N/A - Actually ITG (Machine was synced to an ITG Song)
3. SyncBias=Null (User explicitly set machine to Null and synced to a Null Song)
4. SyncBias=ITG (User explicitly set machine to ITG and synced to a ITG Song)

We can have Song's SyncBias:

1. SyncBias=N/A - Null (Author didn't set, actually null)
2. SyncBias=N/A - ITG (Author didn't set, actually itg)
3. SyncBias=Null (Author marked song as Null offset, preferred)
4. SyncBias=ITG (Author intentionally have added +9ms)

## Combinations to test

For each scenario we should create a test and check calculation properly.
I prepared special pack with `9s` (seconds) sync bias songs to see outcome under test.

The pack is just 4 songs:
- `itg`: offset +9s, SyncBias=ITG
- `n-a-actually-itg`: offset +9s, SyncBias=N/A
- `n-a-actually-null`: offset +0s, SyncBias=N/A
- `null`: offset +0s, SyncBias=Null


| Machine Bias | Song Bias | Test               | Observation                |
|:-------------|:----------|:-------------------|:---------------------------|
| N/A-Null     | N/A-Null  | no-offset-adjust   | Song in sync with machine  |
| N/A-Null     | N/A-ITG   | no-offset-adjust   | Notes earlier, music later |
| N/A-Null     | Null      | no-offset-adjust   | Song in sync with machine  |
| N/A-Null     | ITG       | no-offset-adjust   | Notes earlier, music later |
| N/A-ITG      | N/A-Null  | no-offset-adjust   | Music earlier, notes later |
| N/A-ITG      | N/A-ITG   | no-offset-adjust   | Song in sync with machine  |
| N/A-ITG      | Null      | no-offset-adjust   | Music earlier, notes later |
| N/A-ITG      | ITG       | no-offset-adjust   | Song in sync with machine  |
| Null         | N/A-Null  | no-offset-adjust   | Song in sync with machine  |
| Null         | N/A-ITG   | no-offset-adjust   | Notes earlier, music later |
| Null         | Null      | no-offset-adjust   | Song in sync with machine  |
| Null         | ITG       | itg-on-null-adjust | Song in sync with machine  |
| ITG          | N/A-Null  | no-offset-adjust   | Music earlier, notes later |
| ITG          | N/A-ITG   | no-offset-adjust   | Song in sync with machine  |
| ITG          | Null      | null-on-itg-adjust | Song in sync with machine  |
| ITG          | ITG       | no-offset-adjust   | Song in sync with machine  |

# Global Offset Cases

Global offset contains actual global offset lag of hardware and sound waves,
plus potential itg offset when machine is synced to itg song.

Proper compensate is not possible without sync bias information:

```
# user syncs machine to an itg song:
global-offset = real-hardware-lag + itg-bias

# if rate-music = 2.0
elapsed-time = song-elpased-time + (rate-music * global-offset)

# TODO: I'm still not sure about calculation here
proper-elapsed-time = song-elapsed-time + real-hardware-lag + (itg-bias * rate-music)
```

Here hardware lag is multiplied by rate music and we end up with offsync.

# BUGS

Currently I've been stuck with several blocker bugs:
1. Source Code is cryptic I can't understand where 0BeatOffset must be used as is and where must be used synced version of 0BeatOffset
2. Cache: songs cache calculates `FIRSTSECOND` and `LASTSECOND` on "synced" machine and song. Turning off, for example, machine's sync back to N/A, keeps `offset` untouched, but cached version is calculated on unbiased offset
3. Even without cache sometimes assist tick follows music, sometimes it follows song.

# Results

I've tested all combinations and in general it works for all scenarios. I managed to play in single game on synced machine both ITG and Null biased songs. Old/unmarked songs have offset untouched. When machine is `N/A` everything is untouched and use old behavior.
Tested also rate music - it seems to be working.

And yet, still don't merge.

# Patch extraction

I think I'll extract some refactoring steps in a separate patch which help me to sort out confusing names and code..

@teejusb @sukibaby , I think some review and hints will help me a lot.

sukibaby commented:
I'll do a review today, no worries 🙏

teejusb commented:
What is the long term purpose of this?

IMO the long term goal would be to eliminate the sync bias completely and defaulting to null -- instead of actually formally supporting it long term and adding these fields seems antithetical to long term goals.

I think `group.ini` satisfies the generic labeling for the bias for songs in a pack in a holistic manner, and I think the more beneficial thing to do long term would be something like the following:

1. Define the bias for your local machine (i.e. whether to default NULL or default ITG) via a preference.
2. Add support for generating a `group.ini` for packs that don't have one using that default bias.
3. Add a convenient way to be like "adjust offset for packs that have a differing bias and update the `group.ini` (which is simply just adding/subtracting 9ms to the songs in the pack).

I think the main usage for sync bias in a SSC file is when making compilation/tournament packs. I think pushing for normalization to null and working through the growing pains (which I actually believe will be short lived if we build in tools to assist) is way better long term.

jsirex commented:
@teejusb long term - no, I wish it would as short as possible migration. 
But these options will be as long, as there will be ITG-synced old arcade machines and packs (years).
And for old machines/old packs we can only tag them with `SyncBias`.

We need machine's SyncBias option for long-term:
- ITG: machine mostly have unmarked ITG packs, and it easier to set SyncBias=ITG
- Null: modern machine, owner explicitly synced and set Null.
- N/A: the case when user just upgraded machine, but do not care/aware about new option

I don't agree about `group.ini`, because, SyncBias - is song's specific attribute, not a pack. And yes, packs can have mixed content. I can pull songs from different packs and make "my-mix".
Also, when I put on flash selected songs, I must to mess with syncs and explicitly put on flash `flash-group.ini` and put sync bias. Which is not user friendly. User is not experienced enough to understand such complicated stuff under the hood. User expects, that it is possible to put some songs on flash and play. 

I decided to avoid the case when we touching original file..
We should not touch song's file, because it breaks checksum of the file (I didn't look to the code, is checksum calculated on concrete set of fields, or just file?) and breaks online statistics..
We can't automatically adjust unmarked files to machine's sync bias, because we don't know for sure are these files are actually have that SyncBias. Maybe I loaded bunch of new packs, decided to migrate, but forgot to switch ITG->Null in system settings, but my files already adjusted.. to wrong offset. 

If you ok, I can rewrite logic a bit aggressive pushing towards migration:
1. If machine's SyncBias is set, treat N/A song as it has same SyncBias (but don't touch file)
2. Editor already has support of SyncBias option, but  we can add short cut "Mark N/A song/pack ITG/Null" , but this options should be available in trusted environment only (Home Play, Event Mode). So I would prefer not to do it
3. Forget about old machines, as soon as they've upgraded, treat machine's sync always Null, read song's SyncBias and adjust offset (also, I think way too aggressive, so would prefer not to do it)

sukibaby commented:
> IMO the long term goal would be to eliminate the sync bias completely and defaulting to null -- instead of actually formally supporting it long term and adding these fields seems antithetical to long term goals.

True - **except for the `NotesLoaderSSC` additions which are correct**, and should be implemented for the loaders of other chart variants as well.  As I mentioned in the group.ini PR, we should be factoring for bias within the scope of `NotesLoaderSSC`/`NotesLoaderSM`/etc so that the offset is handled within the domain of the chart itself.

This is more correct than the method in the group.ini PR, because the legacy method of adding 9ms to your simfile's offset value is correctly mitigated when it's handled via `NotesLoader`. This is because, as far as the rest of the engine is concerned, all files it sees are null sync. This would also solve our long-standing issue where offset is scaled incorrectly with changed music rate. In other words, the `NotesLoader` approach fixes most of our problems while allowing us to leave things unchanged in most other places too.

I don't think most of this PR is needed, meanwhile I think the `NotesLoader` changes should apply to other chart formats too.

@jsirex, I understand your point about managing sync on a per-song basis, but to me, that seems like an edge case you could solve by making a folder of 9ms stuff and a folder of 0ms stuff on your USB drive, put the proper group.ini in each, and you'd be able to play the charts correctly on the cab. I think there's too much potential for stuff to get messy going on a per-song basis for managing sync bias. So I apologize, but I agree with @teejusb that it should be on a per-pack/per-folder basis.

sukibaby commented on src/GameSoundManager.cpp:
```diff
@@ -215,7 +215,7 @@ static void StartMusic( MusicToPlay &ToPlay )
 		float fTime = NewMusic->m_NewTiming.GetElapsedTimeFromBeatNoOffset( fDestBeat );
 
-		NewMusic->m_NewTiming.m_fBeat0OffsetInSeconds = fTime;
+		NewMusic->m_NewTiming.SetOffset( fTime );
```
This doesn't need to change

sukibaby commented on src/NotesLoaderSSC.cpp:
```diff
@@ -499,6 +503,9 @@ void SetStepsAttacks(StepsTagInfo& info)
 		info.loader->ProcessAttacks(info.steps->m_Attacks, *info.params);
 	}
 }
+// steps offset/sync bias actually does not make any sense for me
+// there is single audio file and it has single offset
+// implement this anyway
```
So, the problem we're trying to tackle here is when a stepfile's offset value has had 9ms added _after it was synced_ if the author wanted it to have ITG offset.

Don't think about the audio file at all here, what we want to do is detect if the pack is marked as ITG sync, and if so, subtract 9ms from the offset value we read from the file before we give it to the engine.

This way, everything is treated inside the engine as if it's null sync.

jsirex replied:
No, the first problem here simfile format. Why `SetStepsOffset` do exist? Or I just didn't get your comment.

sukibaby commented on src/NotesLoaderSSC.cpp:
```diff
@@ -507,6 +514,15 @@ void SetStepsOffset(StepsTagInfo& info)
 		info.has_own_timing = true;
 	}
 }
+void SetStepsSyncBias(StepsTagInfo& info)
```
This seems overkill, as mentioned above we really just need to subtract the 9ms if we detect the pack has ITG offset. (or add it, if necessary - depends on how this gets implemented:)

> 1. Define the bias for your local machine (i.e. whether to default NULL or default ITG) via a preference.
> 2. Add support for generating a `group.ini` for packs that don't have one using that default bias.
> 3. Add a convenient way to be like "adjust offset for packs that have a differing bias and update the `group.ini` (which is simply just adding/subtracting 9ms to the songs in the pack).

jsirex replied:
Before we can substract 9ms we must somehow know that it is ITG

sukibaby commented on src/NotesLoaderSSC.cpp:
```diff
@@ -998,7 +1019,8 @@ bool SSCLoader::LoadFromSimfile( const RString &sPath, Song &out, bool bFromCach
 					state = GETTING_STEP_INFO;
 					pNewNotes = out.CreateSteps();
-					stepsTiming = TimingData(out.m_SongTiming.m_fBeat0OffsetInSeconds);
+					stepsTiming.m_fBeat0OffsetInSeconds = out.m_SongTiming.m_fBeat0OffsetInSeconds;
```
Let's leave this how it was before too

jsirex replied:
`stepsTiming` instance created at line 991. Creating another instance does not make sense. Am I missing something?

sukibaby commented on src/TimingData.cpp:
```diff
@@ -18,40 +18,75 @@ static const int INVALID_INDEX = -1;
-{
-	/* de-allocate any old pointers we had */
-	Clear();
+void TimingData::Copy(const TimingData &other) {
```
We don't actually need the TimingData changes here, if it's being taken care of within NotesLoader. 

We should factor in the song offset during AdjustSync's autosync feature, however!

jsirex replied:
I see the idea to load song, optionally adjusted and use TimingData with null offset inside. I will think about this.

sukibaby reviewed (commented):
So, as I mentioned in my comment above, most of this is not needed, tbh, as really the changes the NotesLoaderSSC are what I'd want implemented for group.ini. If we go for handling bias on a per-pack/per-folder basis, then a lot of this is unneeded, since it seems like the biggest complications you're running into are due to managing it on a per-song level, which is making the code overcomplicated as a result.

jsirex commented:
@sukibaby 
> @jsirex, I understand your point about managing sync on a per-song basis, but to me, that seems like an edge case you could solve 

No.. ok, I personally can solve this, "a regular user" says "what?" Regular user don't know what is "offset" or ssc file.
And this is one case.

Building mix-packs with existing files - another issue. If, for example, ITL2025 + Unlocks will use different songs from different packs? 

> . I think there's too much potential for stuff to get messy going on a per-song basis

You have to deal with it anyway. Either notes loader just loads a file and reads metadata or notesloader get filename, detect parent folder, check group.ini exists, read it, properly handle file format, handle parse error, then get syncbias.

jsirex commented:
Ok, I had some time to dig code more. 

All your suggestions seem "wiill not work" for me. At least I don't see a clear"how???"

@teejusb :

> I think group.ini satisfies the generic labeling for the bias for songs in a pack

I don't agree SyncBias should be put in goup.ini, but let assume it there...

> Define the bias for your local machine (i.e. whether to default NULL or default ITG) via a preference.

1. What if user does not set it?
2. What if user doesn't know what option to choose?
3. Did you mean that  this option used only as default song's sync bias when song's sync bias is unset?

> Add a convenient way to be like "adjust offset for packs that have a differing bias and update the group.ini (which is simply just adding/subtracting 9ms to the songs in the pack).

1. "Adjust packs that have a differing bias": difference between what and what exactly?
2. Changing anything to the file's content changes it hash and breaks ranking doesn't it?

@sukibaby :

> we should be factoring for bias within the scope of NotesLoaderSSC/NotesLoaderSM/etc so that the offset is handled within the domain of the chart itself.

This will lead to many unpredictable bugs.
If the info about sync bias will be present only during NotesLoader (engine does not aware of it), then:
1. offset is adjusted and put in TimingData
2. cache file will be created with TimingData (adjusted offset to Null, but without syncbias info)
3. First/Last seconds are calculated with Null, used in multiple places (which on ITG machine break couple of things)
4. Preview will be wrong (fortunately no one notice 9ms)
5. Cached TimingData *probably* (it is still complicated) will be used for playing (IDK what will happen when offset in song and cache do not match)
6. AutoSync will also uses TimingData and writes song with wrong (always null) offset (even there is group.ini with ITG sync bias)
7. ... Code is so messed up that IDK what else can go wrong.

99. Potentially, reading offset from cache adjusts offset again according to syncbias and caches new, double-adjusted offset again. Each restart of ITGmania potentially adjusts offset again and again.

>  correctly mitigated when it's handled via NotesLoader

as described above - hardly possible

> changes should apply to other chart formats too.

I decided  to do nothing when it is old file. And support only new files. OLD machine is synced somehow. And for N/A tracks it stays same. For the synced tracks it can appropriately calc offset. 
For any old pack an author can issue a patch with sync bias specified.

> stepfile offset

For me it is still unclear: according to source code `SSC` file can have `OFFSET` tag both in *SONG DATA* section and in *STEPS DATA* section. So I can't imaging a case when `OFFSET` for the same song (same audio file) will be different for different steps/difficulties.

> we really just need to subtract the 9ms if we detect the pack has ITG offset

1.  As mentioned above: will it just break all checksums and rankings?
2. On ITG-synced machine "just subtract" make any ITG-synced songs NULL and breaks things..
3. How can I be sure it hasn't been already subtracted?

# Open questions I need to clarify anyway

Imaging, we release this feature *somehow*. Such release may require additional steps to perform.
We need to decide, is it ok to ask a user to perform a step or we must find another way.

Here is list of steps (some or +wall), potentially required after release:
1. Change machine SyncBias from N/A (default) to explicit NULL/ITG; opt-in approach. (I think, it is ok)
2. Re-sync machine after release (I think, it is ok, and it happened before)
3. Re-sync machine with only using song with SyncBias defined, to properly calculate stuff (Is it too match to ask?)
4. If machine's SyncBias ever changes from Null to ITG or vise versa, the *Cache* folder must be cleaned up (are we ok with that requirements?)
5. Forced upgrade: force machine Null offset, without any options, force adjust any ITG song to Null and require user explicitly mark unmarked ITG-synced song to play them correctly. (Does it look like we're asking waaay to much? Keep in mind, force to change files  == force to change checksums)

What is ok, what is not?


# offtopic

The hash calculated here: https://github.com/itgmania/itgmania/blob/release/src/Song.cpp#L1721
Files are tried in order, but is it the same order to load song?
1. I copy content from SSC to SM file as is
2. I modify SSC file, simplify it having 1 arrow
3. Play on quad
4. Hash matches
5. Score submitted?

jsirex commented:
Close in favor of pack.ini PR
